Release workflow

scottwittenburg opened the pull request:
This topic implements several pieces needed to support a release workflow for Spack.  Some of the pieces included:

1. A new Spack command to create a description of packages for Gitlab-CI to build. It defines a gitlab-ci.yml file that would be placed in a repository, such that when Gitlab-CI detects a commit to the branch, it will build the packages mentioned in the description.  More details on this new command are provided further down in this PR description ("Details of `release-jobs` command).

2. A build job shell script to be run by Gitlab-CI.  The responsibility of this script is to be assigned a spec, determine if the package corresponding to that spec exists and is up-to-date on a configured binary mirror, and if it is not, build the package, create a buildcache for it, and push that buildcache to the binary mirror, communicating with a CDash instance along the way.

3. Other infrastructure and changes in support of the items above.

The command to generate this `.gitlab-ci.yml` file looks like this:

```
spack release-jobs --spec-set <path-to-release.yaml> --mirror-url "https://mirror.spack.io" --shared-runner-tag spack-k8s
```

A summary of the jobs/stages can be printed by the above command if you add the `--print-summary` command-line argument.

### One approach to managing the release workflow

To avoid churn in the Spack repository, we plan not to keep the full list of release jobs under source control in a full `.gitlab-ci.yml` file.  Rather, a separate microservice will be capable of generating the bulk of the contents of the `.gitlab-ci.yml` on demand, for a given commit or sha.  

Here is an example of the `.gitlab-ci.yml` that might be maintained within the Spack repository:

```yaml
include:
  - remote: 'https://internal.spack.io/glciy/${CI_COMMIT_SHA}.yaml'
```

Whenever the pipeline is configured to run, the file above would trigger the microservice to checkout the given sha, run the `spack release-jobs ...` command on the release spec set, and return the full list of staged jobs to be run.

### Details of `release-jobs` command

The `release-jobs` command makes use of two new configuration files defined in the repository, `etc/spack/defaults/release.yaml` (or an arbitrary yaml file following the same schema) and 'share/spack/docker/os-container-mapping.yaml'.  First the `release.yaml` file is used to construct a `CombinatorialSpecSet`, which expands the contents into a list of the entire cross-product of packages/versions/compilers expressed in that file.  Next, the `os-container-mapping.yaml` file is used to find docker containers with compilers that "satisfy" the specs in the expanded cross-product.  The `os-container-mapping.yaml` file is essentially a list of operating systems which each provide an `image` (the image should exist on DockerHub so it can be pulled by both the `release-jobs` command and the AWS gitlab runners), and a list of available compilers on that image.  For any compiler found in one of the containers that satisfies the compiler requirement of a release spec, that spec (with the OS from `os-container-mapping.yaml` added to it) is added to a list that will be concretized in the appropriate docker container in order to gather all the specific dependencies of the spec.  Then all the resulting specs are combined and staged.

For example, given the following `release.yaml`:

```
spec-set:
    include: []
    exclude: []
    matrix:
        - packages:
            libpng:
                versions: [1.6.34]
            ncurses:
                versions: [6.1]
            pkgconf:
                versions: [1.4.2]
            readline:
                versions: [7.0]
            zlib:
                versions: [1.2.11]
        - compilers:
            gcc:
                versions: [5.5.0]
    cdash: ["https://spack.io/cdash/submit.php?project=spack"]
```

and given this `os-container-mapping.yaml`:

```
containers:
  linux-ubuntu18.04-x86_64:
    image: scottwittenburg/spack_builder_ubuntu_18.04
    compilers:
      - name: gcc@5.5.0
      - name: clang@6.0.0-1ubuntu2
  linux-centos7-x86_64:
    image: scottwittenburg/spack_builder_centos_7
    compilers:
      - name: gcc@5.5.0
```

then the `release-jobs` command will generate the following specs to be built:

```
$ spack release-jobs --spec-set .../spack/etc/spack/defaults/release.yaml --mirror-url https://mirror.spack.io --shared-runner-tag spack-k8s --print-summary
==> Staging summary:
==>   stage 0 (6 jobs):
==>     pkgconf/atbmtyn -> pkgconf@1.5.4%gcc@5.5.0 arch=linux-centos7-x86_64
==>     pkgconf/bp4fmns -> pkgconf@1.4.2%gcc@5.5.0 arch=linux-ubuntu18.04-x86_64
==>     pkgconf/qdgufjf -> pkgconf@1.4.2%gcc@5.5.0 arch=linux-centos7-x86_64
==>     pkgconf/unoed43 -> pkgconf@1.5.4%gcc@5.5.0 arch=linux-ubuntu18.04-x86_64
==>     zlib/ijtgxbh -> zlib@1.2.11%gcc@5.5.0 arch=linux-centos7-x86_64
==>     zlib/ymics2i -> zlib@1.2.11%gcc@5.5.0 arch=linux-ubuntu18.04-x86_64
==>   stage 1 (4 jobs):
==>     libpng/2vicq3d -> libpng@1.6.34%gcc@5.5.0 arch=linux-ubuntu18.04-x86_64
==>     libpng/drgxai6 -> libpng@1.6.34%gcc@5.5.0 arch=linux-centos7-x86_64
==>     ncurses/aqchhtc -> ncurses@6.1%gcc@5.5.0 arch=linux-centos7-x86_64
==>     ncurses/qkjccof -> ncurses@6.1%gcc@5.5.0 arch=linux-ubuntu18.04-x86_64
==>   stage 2 (2 jobs):
==>     readline/vuigree -> readline@7.0%gcc@5.5.0 arch=linux-ubuntu18.04-x86_64
==>     readline/xlb6j2a -> readline@7.0%gcc@5.5.0 arch=linux-centos7-x86_64
==> 12 build jobs generated in 3 stages
```

Note that although the `clang@6.0.0-1ubuntu2` compiler is available in the `linux-ubuntu18.04-x86_64` container, no jobs with that compiler were staged because it wasn't listed as a desired compiler in the `release.yaml`.  Conversely, also note the additional version of `pkgconf` which will be built, even though it wasn't specifically mentioned in the `release.yaml`.  This is because that particular version was a dependency of another spec in the spec set.

To avoid running the concretization of release specs (and generation of dependencies) in docker containers, the `--this-machine-only` argument may be passed to the `release-jobs` command.  In this case, the current machine will be used for concretization and dependency generation, but for this to work, the `os-container-mapping.yaml` file must contain an entry corresponding to `spack.architecture.sys_type().

### Additional Notes

Currently the Spack Gitlab instance is set up to test branches it finds on my fork (`scottwittenburg/spack`), so that will need to be changed before merging this topic will seem to have any effect.

This PR replaces both #8451 and #8718.

Before we can merge this, we need to address at least #10141, but there may be other prerequisites as well, hence the `WIP` on this PR for the time being.

Things currently holding this up:

- [x] Fix issue installing from spec.yaml using `-f` #10141 
- [x] Fix issue concretizing xsdk-0.4.0 (just needs merge) #10288 
- [x] Fix omega_h regression (proposed fix here: #10297)

scottwittenburg commented:
@gartung I don't think it's necessarily precisely what we want yet, but I think we'd like to get something merged and then iterate on it from there.  I think this version is pretty close to something we could merge though.  Maybe you could take a look and provide your thoughts?  Let me know if you have questions or want something clarified.  Pinging @tgamblin as he'll probably be interested in your thoughts as well.

gartung commented:
The idea of spec sets is appealing. We need the ability to specify a preferred set of package versions that may not be the latest. Would the spec set package versions be used in concretization?

scottwittenburg commented:
@gartung Yes, the `CombinatorialSpecSet` will expand the `release.yaml` out to the full cross-product of packages x versions x compilers specified there.  Currently each spec in the resulting list of specs is then concretized in any docker containers specified in the `os-container-mapping.yaml` where there is a compiler that satisfies the one in the spec.  And then finally each of those (a specific pkg/version/compiler/os) becomes a job/package to be built by the Gitlab instance.  The jobs are staged in such a way that for any package to be built, Gitlab has access to it's dependencies in binary form, this reduces the amount of work to be done and speeds up the builds.

Down the road I think @tgamblin would like to avoid needing to concretize each spec in a specific container in order to generate the `.gitlab-ci.yml`, but that's how we do things at the moment.

Hopefully I answered your question, but let me know if something is unclear.

gartung commented:
Have you considered the case where you want to specify variants and compiler flags in the spec-set as well?

scottwittenburg commented:
Yes, we've talked about that, and would like to get there eventually, but there's no support for it here yet.

gartung commented:
The only way you can enforce the version, variants and flags is to manage spack/etc/spack/packages.yaml. This was problematic with different release having a different packages.yaml from the current one. The environments feature helps with this because it preserves a snapshot of the concretization.

gartung commented:
I have determined that the buildcache creation with placeholder rpaths is bugged. The original install path is supposed to be replaced with a placeholder in the rpath list. Instead the path with the placeholder is added to the rpath list. This is why the allow root flag always has to be provided.

gartung commented:
I take that back. The bug appears to be checking the first path in rpaths for the placeholder string. This turns out to be the compiler path for c++ libraries.

gartung commented:
See https://github.com/spack/spack/pull/10287 for a fix to the problem I described.

tgamblin commented:
@gartung: have you seen the new [environments support](https://spack.readthedocs.io/en/latest/tutorial_environments.html)?  We have some features in the pipeline that will build on it for better spec sets -- particularly for combinatorial builds.  Maybe we should chat Thursday about the work we are doing on Spack Stacks, which will basically be combinatorial environments -- "instances" of combinatorial deployments.  Look at https://github.com/spack/spack/projects/7.  Stacks will likely to be a good long-term way to specify a release and they'll be implemented using some of this.

scottwittenburg commented:
@tgamblin @balay The branch associated with this PR is being tested on our AWS infrastructure (pipeline [176](https://gitlab.spack.io/scott/spack/pipelines/176)), and currently it's hitting a failure on both centos 7 and ubuntu 18.04 during the configuration phase of `omega_h`.  I've created an [issue](https://github.com/ibaned/omega_h/issues/282) there to get some help resolving it.

balay commented:
I've filed  omega-h issue at https://github.com/spack/spack/issues/10291

scottwittenburg commented:
ping @opadron

scottwittenburg commented:
ping @aashish24

scottwittenburg commented:
@tgamblin I have removed the `.gitlab-ci.yml` since we're in the process of creating a microservice in the kubernetes cluster to do that.  Here's the last pipeline that ran on this branch before I removed it though:

https://gitlab.spack.io/scott/spack/pipelines/179

scottwittenburg commented:
I'm not sure why github is still showing the file in the diff...

scottwittenburg commented on lib/spack/spack/util/spec_set.py:
```diff
@@ -0,0 +1,189 @@
+from llnl.util.tty.colify import colify
+
+import spack
+import spack.compilers
```
I took this file from a commit by @tgamblin initially, and have no argument against moving it.  So I'll do it, it's easy enough to move back if there was a good reason for it to leave in `spack.util`.

tgamblin replied:
@scheibelp: I think it's ok for stuff in `spack.util` to use core... sparingly.  The stuff in `llnl.util` definitely shouldn't have any lines that import from the `spack` package, but `spack.util` is *supposed* to be stuff that is sort of ancillary to the core but can use it.  So far these packages import from spack core:

* `executable.py`
* `gpg.py`
* `lock.py`
* `naming.py`
* `package_hash.py`
* `path.py`
* `spack_json.py`
* `spack_yaml.py`
* `web.py`

It's a blurry distinction, though and we've never really defined it clearly.

tgamblin replied:
so I don't think this has to be moved, unless @scheibelp has a preference

scheibelp replied:
Amongst all the listed modules in `spack/util` I think that only `package_hash` does "significant" imports from the core. That is certainly an exception though. Perhaps they should both live in `lib/spack`. That isn't essential to sort out here though so I'll consider this resolved for now.

scottwittenburg commented on lib/spack/spack/cmd/release_jobs.py:
```diff
@@ -0,0 +1,406 @@
+
+def get_spec_string(spec):
+    if spec.architecture:
+        return '{0}@{1}%{2} arch={3}'.format(spec.name, spec.version,
```
Is the only difference going to be that I'll get all the variants added if I use `spec.format()` here?  I didn't want it to get too verbose.

scottwittenburg commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -645,3 +696,147 @@ def get_keys(install=False, trust=False, force=False):
+
+        return True
+
+    except Exception as e:
```
Here I was intending that any other kind of exception that I didn't expect should trigger the package to be rebuilt, and failing through would prevent the package from getting rebuilt.  Is there a better way to catch any unforeseen exceptions?

scottwittenburg commented on lib/spack/spack/cmd/release_jobs.py:
```diff
@@ -0,0 +1,406 @@
+        help="path to release spec-set yaml file")
+
+    subparser.add_argument(
+        '-m', '--mirror-url', default='http://172.17.0.1:8081/',
```
Hmmm, I kind of disagree.  This was definitely an old one from early testing days, and the one I use now is `https://mirror.spack.io`.  I'm replacing this old default with the new one for now, but we can iterate if you feel strongly we should not have any default.

scheibelp commented on lib/spack/spack/cmd/release_jobs.py:
```diff
@@ -0,0 +1,406 @@
+        help="The spec for which you want container-generated deps")
+
+    subparser.add_argument(
+        '--this-machine-only', action='store_true', default=False,
```
I think a clearer option name would be `--concretize-locally` or `--resolve-deps-locally`, since this just determines the dependencies locally (the description of the option is informative but the option could be named more-suggestively)

scottwittenburg commented:
@scheibelp Thanks for the review.  I have addressed many of the concerns you brought up, most significantly how the specs are concretized in the docker containers and that information is communicated back to the originating process/command.  I still have a few of your comments/suggestions to address, but I won't be able to get to them until next week.  Once I think I've hit everything, I'll let you know.  In the meantime, feel free to start looking at the things I *have* managed to address.

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -21,12 +25,19 @@
 
 
+_build_cache_relative_path = 'build_cache'
+_meta_yaml_regex = re.compile(
```
Is this still needed? I don't see it being used

scottwittenburg commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -366,7 +417,7 @@ def download_tarball(spec):
     tarball = tarball_path_name(spec, '.spack')
     for key in mirrors:
-        url = mirrors[key] + "/build_cache/" + tarball
+        url = mirrors[key] + "/" + _build_cache_relative_path + "/" + tarball
```
I'm fixing the `for key in mirrors` bit, but since this is a url, should we just continue to build it manually using the `... + "/" + ...` approach?

scottwittenburg commented:
@scheibelp Do you have an idea where you think the `release-jobs` command description I added to the PR description should live?  I was poking around in the `lib/spack/docs/` folder and it didn't seem obvious how best to organize this new bit of documentation.

scottwittenburg commented on lib/spack/spack/cmd/release_jobs.py:
```diff
@@ -0,0 +1,486 @@
+
+        for s in spec.traverse(deptype=deptype):
+            if not s.concrete:
+                s.concretize()
```
I'm not sure I follow.  Is it true that `s.traverse()` is generating the dependency specs of `spec`?  Until I have my hands on one of those dependency specs, how I can check if it's concrete?  I experimented with this code a bit to try and figure out how it worked, but I basically stole it from [here](https://github.com/spack/spack/blob/ec78d362d21d420fe6ee24bf888cb6fc2af37611/lib/spack/spack/graph.py#L562).  I added the concretize() in there myself though, because I wasn't getting all the dependencies without it.  Maybe we can chat about this one at some point if you have time.

scheibelp replied:
If X->Y->Z, and you concretize X with `X.concretize()`, all dependencies of X will also be concrete (that's part of the API of `X.concretize()` - we should update the docstring if that is unclear). If you think dependencies are missing from X after calling `X.concretize()` that is a bug in Spack.

scottwittenburg replied:
Ok, thanks for the clarification.  I'll find out for sure what's going on there, and file an issue if needed.

scottwittenburg replied:
I did some testing here, and it seems there was no problem removing the `concretize()` from within the `traverse()`, as you suggested.  So that's gone now, thanks.

scheibelp replied:
I still see that logic here, has it not been pushed yet?

scottwittenburg replied:
Ah for pete's sake.  Maybe a `git` mistake on my part?  I'll fix it, sorry about that.

scottwittenburg commented on lib/spack/spack/cmd/release_jobs.py:
```diff
@@ -0,0 +1,486 @@
+        next_stage = unstaged.difference(depends_on)
+        stages.append(next_stage)
+        unstaged.difference_update(next_stage)
+        deps = remove_satisfied_deps(deps, next_stage)
```
I like that wording, hope you don't mind if I borrow it.

scottwittenburg commented on lib/spack/spack/cmd/release_jobs.py:
```diff
@@ -0,0 +1,486 @@
+            specs = container_specs[osname]['specs']
+            get_spec_dependencies(specs, deps, spec_labels, image)
+
+    dependencies = deps
```
After a little consideration, I felt the best thing to do was to add the comment you mentioned and then to work with `dependencies` in the loop, returning `deps`.  But see what you think.

scottwittenburg commented:
@scheibelp Thanks for providing some thoughts on how to organize some documentation for this.  I'd like to take a stab at that, but what do you think about reserving that for the next PR?  If you think it's important to get it in this one, I'm fine with that.

scheibelp commented:
> I'd like to take a stab at [documentation], but what do you think about reserving that for the next PR?

That works for me. In your opinion is this ready for another look?

scottwittenburg commented:
Yes I think so, thanks @scheibelp.

scottwittenburg commented:
@scheibelp When I finally got Travis to pass all the tests this morning, I noticed the targeted coverage of new code is quite low, so I'm in the process of adding a couple tests of the `release_jobs.py` module.

scheibelp commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -106,7 +193,9 @@ def find_matching_specs(
     has_errors = False
+    tty.msg('find_matching_specs: going to parse specs')
     specs = spack.cmd.parse_specs(pkgs)
+    tty.msg('find_matching_specs: going to parse specs')
```
Appears to be a duplicate of two lines prior. Also was this originally added for debugging purposes? My guess is there should be one call and it should be `tty.debug`

scheibelp commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -272,6 +380,163 @@ def getkeys(args):
+        root_spec.concretize()
+        spec = root_spec
+        for dep_spec in args.specs.split():
+            # if args.root_spec != dep_spec:
```
These commented lines can be removed I assume (here and the `concretize` just below)

scheibelp commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -272,6 +380,163 @@ def getkeys(args):
+    sys.exit(0)
+
+
+def save_dependent_spec_yaml(args):
```
IMO this would be a good utility function, not just something available to `build_cache`. Also, I think it would be better if it read in a `.yaml` spec file (so that you didn't have to specify the root), and extracted the dependency yaml from that.

scottwittenburg replied:
In that case, do you think it should live in `lib/spack/spack/spec.py`?  Should it be a module-level method or a static member of `Spec`?  And regardless of that decision, what do you think of arguments like:

```python
def save_dependency_spec_yamls(root_spec_as_yaml, output_directory, dependencies=None)
```

And the caller would pass `dependencies` (a list of spec names) if they want filter out which dependencies get written, otherwise, it would write out a `.spec.yaml` for every dependency.  This would be convenient for my build jobs, which would rather just call one spack command and get all the full specs it cares about written to disk at once.

Also note that my build jobs don't have their hands on the root spec yaml file, and need to concretize the abstract root spec themselves, and so it might be nice if this method could take the root spec as a yaml object rather than insisting on reading it from a file.

But let me know what you think.

scheibelp replied:
> In that case, do you think it should live in lib/spack/spack/spec.py

That sounds good

> Should it be a module-level method or a static member of Spec

Either is fine

> what do you think of arguments like:

Those arguments look good

> note that my build jobs don't have their hands on the root spec yaml file, and need to concretize the abstract root spec themselves, and so it might be nice if this method could take the root spec as a yaml object rather than insisting on reading it from a file.

That also works

scheibelp commented on bin/rebuild-package.sh:
```diff
@@ -0,0 +1,311 @@
+    #  - the log file will be passed in as the first argument, and
+    #  - the code in this function will not be logged.
+    echo "The full log file is located at $1"
+    # TODO: send this log data to cdash!
```
Is this still a TODO? Is this pending some other thing being done?

scottwittenburg replied:
I have a refactoring pass planned for the rebuild script.  One of the things I want to do there is to take control of job registration with CDash (instead of letting `spack install ...` do it).  This way, any errors encountered in the bash script can be potentially included in the information given to CDash and available for viewing there.  As it stands, nothing which may go wrong before the `spack install...` command could possibly be communicated to CDash.  I think it would also be nice if the full output log produced by this rebuild script is pushed to CDash at that point as a "note".  But I thought we wanted to get this topic merged as quickly as possible, so I didn't want to hold it up with a bunch of things that could be done as follow-on topics, hence that `TODO` comment.

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -235,35 +250,50 @@ def sign_tarball(key, force, specfile_path):
 
-def generate_index(outdir, indexfile_path):
-    f = open(indexfile_path, 'w')
+def generate_html_index(path_list, output_path):
```
Looks like this is primarily a helper function for `generate_package_index`. In that case could you indicate that and prefix it with `_`, like `_generate_html_index`?

scottwittenburg commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -645,3 +681,131 @@ def get_keys(install=False, trust=False, force=False):
+    spec_yaml_file_name = tarball_name(spec, '.spec.yaml')
+    file_path = os.path.join(build_cache_dir, spec_yaml_file_name)
+
+    result_of_error = 'Package ({0}) will {1}be rebuilt'.format(
```
If you don't like how the code looks, I'm happy to change it.  However, the string I format with may be the empty string or else it may be `"not "`.   So there shouldn't be a missing space in the log/error message.

scheibelp replied:
Got it - I overlooked that.

scottwittenburg commented:
> I had a question regarding
> 
> > This topic implements the release workflow for automatically triggering the Spack Gitlab instance to build a set of "release" packages and push them to the binary mirror on S3.
> 
> what triggers the gitlab instance to build the packages?

Currently, the gitlab instance is configured to test any branch pushed to my fork of spack.  Of course, it won't do anything with branches that don't contain a `.gitlab-ci.yml` at the root.  So concurrently with merging this topic (or shortly thereafter), the plan is to change that gitlab configuration so that 1) it is linked to the main spack repo instead of my fork, and 2) it only triggers the pipeline when a certain branch is pushed (maybe `release`, but I forget exactly which branch(es) we discussed).

scheibelp commented on bin/rebuild-package.sh:
```diff
@@ -0,0 +1,320 @@
+### This script represents a gitlab-ci job, corresponding to a single release
+### spec.  As such this script must first decide whether or not the spec it
+### has been assigned is up to date on the remote binary mirror.  If it is
+### not (i.e. the source code has changed in a "significant enough" way), this
```
Since it's using the full package hash, this would be if the source code changed at all, wouldn't it?

scottwittenburg commented on lib/spack/spack/cmd/release_jobs.py:
```diff
@@ -0,0 +1,568 @@
+        # Note that "dependencies" is a dictionary mapping each dependent
+        # package to the set of not-yet-handled dependencies.  The final step
+        # below removes all the dependencies that are handled by this stage.
+        # Note that in the line the below, the "dependencies" variable is
```
I see, makes sense.

scottwittenburg commented:
> > Currently, the gitlab instance is configured to test any branch pushed to my fork of spack.
> 
> Is that to say, any branch, or any commit to that branch?

The way it is configured now, any time any branch is pushed, the gitlab instance will discover it eventually (within 30 minutes at most I think) and run the jobs in the `.gitlab-ci.yml`.

> It seems that rather than triggering Gitlab to build the packages, this defines the `gitlab-ci.yml` that is used when Gitlab detects a new branch (or commit). If that seems reasonable, then I think this part of the description should be reworded:

Sounds good, will do.

> > This topic implements the release workflow for automatically triggering the Spack Gitlab instance to build a set of "release" packages and push them to the binary mirror on S3.

scheibelp commented:
> This topic implements the release workflow for automatically triggering the Spack Gitlab instance to build a set of "release" packages and push them to the binary mirror on S3.

Ah sorry I was not clear: what I dislike about this explanation is that it implies that some part of what is added to this PR triggers Gitlab, when it is actually just pushing to the branch that triggers Gitlab. Does something like

> This adds a command to create a description of packages for Gitlab to build. It defines a gitlab-ci.yml file that would be placed in a repository, such that when Gitlab detects a commit to the branch, it will build the packages mentioned in the description.

sound alright?

scottwittenburg commented:
@scheibelp I have updated the PR description a bit, to include the text you suggested just above.  Have a look and let me know what you think.

scheibelp commented on bin/rebuild-package.sh:
```diff
@@ -0,0 +1,320 @@
+### This script represents a gitlab-ci job, corresponding to a single release
+### spec.  As such this script must first decide whether or not the spec it
+### has been assigned is up to date on the remote binary mirror.  If it is
+### not (i.e. the source code has changed in a way that caused a change in the
```
Continuing https://github.com/spack/spack/pull/10274#discussion_r253725186

If the source code changes by one character, that will change the hash of the archive (we specifically use hash algorithms that strive for this property), which is part of the full hash. Therefore I think it is clearer to say:

> if the source code has changed

and exclude "in a way that...". Since all changes should produce a different hash.

scheibelp commented:
> I have updated the PR description a bit, to include the text you suggested just above. Have a look and let me know what you think.

The new description looks good - thanks for updating it!

scottwittenburg commented on lib/spack/spack/test/spec_set.py:
```diff
@@ -0,0 +1,286 @@
+        'spec-set': {
+            'include': ['gmake', 'appres'],
+            'matrix': [
+                {'specs': [
```
I've added comments to the tests, hopefully the intent of each one is more clear now.

scottwittenburg commented on lib/spack/spack/test/spec_set.py:
```diff
@@ -0,0 +1,286 @@
+    assert len(specs) == 8
+
+
+def test_spec_set_compilers_bad_property():
```
Agreed, I removed this test, since just constructing a `CombinatorialSpecSet` validates the yaml.

scottwittenburg commented on lib/spack/spack/test/spec_set.py:
```diff
@@ -0,0 +1,286 @@
+        CombinatorialSpecSet(yaml_file)
+
+
+def test_spec_set_get_cdash_string():
```
Makes sense, I did this.

scottwittenburg commented on lib/spack/spack/spec_set.py:
```diff
@@ -0,0 +1,189 @@
+
+        Args:
+            yaml_like: either raw YAML data as a dict, a file-like object
+                to read the YAML from, or a string containing YAML.  * A
```
I think that may just have been a typo, I removed that duplicated sentence.

scottwittenburg commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -216,8 +321,10 @@ def createtarball(args):
 
     for spec in specs:
         tty.msg('creating binary cache file for package %s ' % spec.format())
+        spec.concretize()
```
Sounds good, I have removed that `concretize()` call.

scottwittenburg commented on lib/spack/docs/example_files/spec_set.yaml:
```diff
@@ -0,0 +1,22 @@
+---
```
Oversight.  Removed.

scottwittenburg commented on etc/spack/defaults/release.yaml:
```diff
@@ -0,0 +1,26 @@
+# -------------------------------------------------------------------------
+# This is the default spack release package list file.
```
Fixed.

scottwittenburg commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -645,3 +681,131 @@ def get_keys(install=False, trust=False, force=False):
+
+
+def check_specs_against_mirrors(mirrors, specs, output_file=None,
+                                rebuild_on_errors=False):
```
Done.

scottwittenburg commented on share/spack/docker/spack_builder/update_rpaths.py:
```diff
@@ -0,0 +1,40 @@
+
+
+def update_compiler(compilers_path, prefix, rpaths):
+    with open(compilers_path, 'r') as fd:
```
I guess I just wasn't aware enough of the functionality.  Thanks for pointing it out, I have changed this code to use Spack's config parser.  I also changed the Dockerfile which uses this and rebuilt the Centos7 image.

scottwittenburg commented:
@scheibelp @tgamblin When we get to where we're happy with the code in this PR, I'll squash all the review-related changes into a single commit.

scheibelp commented:
As I understand, originally the intent was to use `spack release-jobs` to generate a `gitlab-ci.yml` that would be committed to the Spack repository and would define all the packages that would be built; now, the goal is to have a `gitlab-ci.yml' which points to a location where Gitlab can pull the full set of specs to build (and `spack release-jobs` would define the `gitlab-ci.yml` file that the Spack repository points to outside of Spack).

If that's correct, I was assuming the "pointer" `gitlab-ci.yml` file would be here, but I don't see it. Am I missing something?

scottwittenburg commented:
No, I don't think you're missing anything.  At the moment we're thinking we'll have another small PR once this one is merged and do a few things all at once at that time (add that "pointer" `.gitlab-ci.yml`, change the gitlab instance configuration so it builds only certain branch(es) from the main repo, and a few other things).  

Take a look at this [branch](https://github.com/scottwittenburg/spack/commits/release-workflow-testing) on my fork to see an example of what that "pointer" `.gitlab-ci.yml` will look like (though the "included" url will have a CI variable in it which will point to the SHA the microservice should check out before running `spack release-jobs ...`).  You can see the resulting pipeline from that testing branch on our "staging" gitlab [here](https://gitlab.staging.spack.io/scott/spack/pipelines/224).  

Let me know what you think.  Or if my explanation wasn't very clear.

scheibelp commented on lib/spack/spack/test/spec_set.py:
```diff
@@ -0,0 +1,298 @@
+    assert Spec('%gcc@4.2.1') in compilers
+    assert Spec('%gcc@6.3.0') in compilers
+    assert Spec('%clang@8.0') in compilers
+    assert Spec('%clang@3.8') in compilers
```
perhaps also assert that `len(compilers) == 4`

scottwittenburg replied:
Added that check.

scottwittenburg commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -88,6 +100,81 @@ def setup_parser(subparser):
+    saveyaml.add_argument(
+        '-r', '--root-spec', default=None,
+        help='Root spec of dependent spec')
+    saveyaml.add_argument(
```
Just to be sure I understand you: you're suggesting it should be a subcommand in `lib/spack/spack/cmd/spec.py`?  Hmmm, yeah, I do agree with you if that's the case.  I probably should have thought of that when you asked me to move the actual functionality that implements the command into the `spack.spec` module.  Suddenly it looks a little out of place here.

scottwittenburg replied:
But yes, let's hold off on this for the next PR.

scottwittenburg commented on bin/rebuild-package.sh:
```diff
@@ -0,0 +1,320 @@
+    spack mirror add remote_binary_mirror ${MIRROR_URL}
+
+    # Now download it
+    spack -d buildcache download --spec-yaml "${SPEC_YAML_PATH}" --path "${BUILD_CACHE_DIR}/" --require-cdashid
```
So I did remove understanding of CDash from the `binary_distribution.py` module, as you requested in a previous round of change requests.  But at the time I didn't see how to remove it from here without forcing the caller (the user entering the command in the terminal) to understand how the buildcache entry files are named.  But maybe it's as simple as just grabbing any files with the same path and prefix as the `.spec.yaml` file, but in that case, we'd need to download some kind of index of all files in the root of the build_cache directory to know what those were.

I hadn't thought of this as necessarily related to the CDash-awareness of `spack install`, except at a higher conceptual level.  But maybe it is.  Though you're not suggesting we re-work the way reporters are hooked into the install process in general, are you?

scottwittenburg replied:
Didn't do anything here, as we covered this earlier today in conversation.

scottwittenburg commented on bin/rebuild-package.sh:
```diff
@@ -0,0 +1,320 @@
+    # Now get CDash ids for dependencies and "relate" each dependency build
+    # with this jobs build
+    for DEP_PKG_NAME in "${JOB_DEPS_PKG_NAMES[@]}"; do
+        echo "Getting cdash id for dependency --> ${i} <--"
```
Wow, good eye.  No, you're not missing anything about bash syntax, that's just a bug (albeit only in a print statement).  I'll fix that tomorrow morning when I address the rest of these comments.  Thanks!

scottwittenburg commented on bin/rebuild-package.sh:
```diff
@@ -0,0 +1,320 @@
+
+# Show the compiler details
+echo "Compiler Configurations:"
+cat ~/.spack/linux/compilers.yaml
```
Ah, nice, I was unaware of that.  I'll make that change.

scottwittenburg commented:
Thanks for another careful round of review and more thoughtful comments @scheibelp, I should be able to knock these all out before we talk tomorrow.

scottwittenburg commented on lib/spack/spack/cmd/release_jobs.py:
```diff
@@ -0,0 +1,570 @@
+    if unstaged:
+        stages.append(unstaged.copy())
+
+    return spec_labels, deps, stages
```
Added a long docstring here, hopefully it's not too much.

scheibelp commented:
Looks good: thanks! Once you have refactored the commits, let me know and I can browse them and merge this.

scottwittenburg commented:
Thanks for all the hard work reviewing this beast @scheibelp.  I have force-pushed the rewritten history, so have a look at the commit messages when you get a chance.  Those will be easy enough for me to update if you want changes.

scheibelp commented:
Thanks!

gartung commented:
This error was produced when trying to create a buildcache for gcc@7.3.0 that was linked against system zlib which has an entry for a version 1.2.7 that does not exists in the zliz package.py

```
  zlib:
    paths:
      zlib@1.2.7+optimize+pic+shared arch=linux-rhel7-x86_64: /usr
```

```
==> [2019-02-27-20:52:54.622705] Warning: Missing a source id for zlib@1.2.7
Traceback (most recent call last):
  File "/build/work/spack-tools/spack/bin/spack", line 48, in <module>
    sys.exit(spack.main.main())
  File "/build/work/spack-tools/spack/lib/spack/spack/main.py", line 675, in main
    return _invoke_command(command, parser, args, unknown)
  File "/build/work/spack-tools/spack/lib/spack/spack/main.py", line 446, in _invoke_command
    return_val = command(parser, args)
  File "/build/work/spack-tools/spack/lib/spack/spack/cmd/buildcache.py", line 528, in buildcache
    args.func(args)
  File "/build/work/spack-tools/spack/lib/spack/spack/cmd/buildcache.py", line 326, in createtarball
    not args.no_rebuild_index)
  File "/build/work/spack-tools/spack/lib/spack/spack/binary_distribution.py", line 363, in build_tarball
    spec_dict['full_hash'] = spec.full_hash()
  File "/build/work/spack-tools/spack/lib/spack/spack/spec.py", line 1306, in full_hash
    package_hash = self.package.content_hash()
  File "/build/work/spack-tools/spack/lib/spack/spack/package.py", line 1092, in content_hash
    for p in self.spec.patches)
  File "/build/work/spack-tools/spack/lib/spack/llnl/util/lang.py", line 183, in __call__
    self.cache[args] = self.func(*args)
  File "/build/work/spack-tools/spack/lib/spack/spack/spec.py", line 2690, in patches
    for sha256 in self.variants['patches']._patches_in_order_of_appearance:
AttributeError: 'MultiValuedVariant' object has no attribute '_patches_in_order_of_appearance'
```

gartung commented:
@scottwittenburg it appears to happen while calculating spec.full_hash()

scottwittenburg commented:
@scheibelp I had a fix for this in my branch at one time, but took it out after some discussion that it was not required anymore.  Should I dig it up and create a PR?  It was just a guard against using that attribute if it didn't exist.

gartung commented:
The fix is trivial.
```
diff --git a/lib/spack/spack/spec.py b/lib/spack/spack/spec.py
index 76c06d3..2b132fa 100644
--- a/lib/spack/spack/spec.py
+++ b/lib/spack/spack/spec.py
@@ -2678,6 +2678,10 @@ class Spec(object):
         if not self.concrete:
             raise SpecError("Spec is not concrete: " + str(self))
 
+        if not hasattr(self.variants['patches'],
+                       '_patches_in_order_of_appearance'):
+            return []
+
         if 'patches' not in self.variants:
             return []
```

scottwittenburg commented:
@gartung Indeed, that's what I had in my branch, but I removed because after discussing it with @scheibelp, it seemed that should not be needed.

gartung commented:
It's a weird corner case where the package was built before  
variants['patches']._patches_in_order_of_appearance
was added.

I have other packages built in another (newer) spack directory where the same zlib@1.2.7 was used in packages.yaml and packages built against it do not give this error when creating the buildcache.

I added this patch to my current buildcache-placeholder-fix branch. We'll see if it passes review.

scottwittenburg commented:
> It's a weird corner case where the package was built before
> variants['patches']._patches_in_order_of_appearance
> was added.

Ah, then it makes sense why I seemed to need it in my local testing, but in general it seems like it shouldn't be necessary.

> I have other packages built in another (newer) spack directory where the same zlib@1.2.7 was used in packages.yaml and packages built against it do not give this error when creating the buildcache.
> 
> I added this patch to my current buildcache-placeholder-fix branch. We'll see if it passes review.

Perhaps a link to this conversation from your current PR would help the reviewer understand why that guard is there.

Thanks for this.
